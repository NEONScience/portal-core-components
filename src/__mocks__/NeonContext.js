/**
   NeonContext Mock

   Various tests cover modules that use NeonContext to get state/site/domain metadata. By mocking
   NeonContext and plugging in all fallback data the tested modules will be able to use a hydrated
   NeonContext for more complete testability.

   Usage:
   import 'path/to/src/__mocks__/NeonContext';
*/

import sitesJSON from '../sampleData/sites.json';
import bundlesJSON from '../sampleData/bundles.json';
import statesJSON from '../lib_components/staticJSON/states.json';
import domainsJSON from '../lib_components/staticJSON/domains.json';
import timeSeriesDataProductsJSON from '../lib_components/staticJSON/timeSeriesDataProducts.json';

import NeonContext from '../lib_components/components/NeonContext/NeonContext';
import BundleParser from '../lib_components/parser/BundleParser';

jest.mock('../lib_components/components/NeonContext/NeonContext', () => (
  {
    ...(jest.requireActual('../lib_components/components/NeonContext/NeonContext').default),
    useNeonContextState: jest.fn(),
    FETCH_STATUS: {
      AWAITING_CALL: 'AWAITING_CALL',
      FETCHING: 'FETCHING',
      ERROR: 'ERROR',
      SUCCESS: 'SUCCESS',
    },
  }
));

NeonContext.useNeonContextState.mockReturnValue([
  {
    ...NeonContext.DEFAULT_STATE,
    data: {
      sites: sitesJSON,
      states: statesJSON,
      domains: domainsJSON,
      bundles: BundleParser.parseContext(
        BundleParser.parseBundlesResponse(bundlesJSON),
      ),
      timeSeriesDataProducts: timeSeriesDataProductsJSON,
      stateSites: {}, // derived when sites is fetched, needs to be mocked here?
      domainSites: {}, // derived when sites is fetched, needs to be mocked here?
    },
    isFinal: true,
  },
]);
