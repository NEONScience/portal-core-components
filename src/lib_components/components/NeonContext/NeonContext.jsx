import React, {
  createContext,
  useContext,
  useReducer,
  useEffect,
} from 'react';
import PropTypes from 'prop-types';

import cloneDeep from 'lodash/cloneDeep';

import { of } from 'rxjs';
import { map, catchError } from 'rxjs/operators';

import REMOTE_ASSETS from '../../remoteAssetsMap/remoteAssetsMap';
import AuthService from '../NeonAuth/AuthService';
import NeonGraphQL from '../NeonGraphQL/NeonGraphQL';
import sitesJSON from '../../staticJSON/sites.json';
import statesJSON from '../../staticJSON/states.json';
import domainsJSON from '../../staticJSON/domains.json';
import bundlesJSON from '../../staticJSON/bundles.json';
import timeSeriesDataProductsJSON from '../../staticJSON/timeSeriesDataProducts.json';

const DRUPAL_HEADER_HTML = REMOTE_ASSETS.DRUPAL_HEADER_HTML.KEY;
const DRUPAL_FOOTER_HTML = REMOTE_ASSETS.DRUPAL_FOOTER_HTML.KEY;

export const FETCH_STATUS = {
  AWAITING_CALL: 'AWAITING_CALL',
  FETCHING: 'FETCHING',
  ERROR: 'ERROR',
  SUCCESS: 'SUCCESS',
};

const DEFAULT_STATE = {
  data: {
    sites: {},
    states: statesJSON,
    domains: domainsJSON,
    bundles: bundlesJSON,
    timeSeriesDataProducts: timeSeriesDataProductsJSON,
    stateSites: {}, // derived when sites is fetched
    domainSites: {}, // derived when sites is fetched
  },
  html: {
    [DRUPAL_HEADER_HTML]: null,
    [DRUPAL_FOOTER_HTML]: null,
  },
  fetches: {
    sites: { status: FETCH_STATUS.AWAITING_CALL, error: null },
    auth: { status: null, error: null },
    [DRUPAL_HEADER_HTML]: { status: null, error: null },
    [DRUPAL_FOOTER_HTML]: { status: null, error: null },
  },
  auth: {
    useCore: false,
    isAuthenticated: false,
    isAuthWorking: false,
    isAuthWsConnected: false,
    userData: null,
  },
  isActive: false,
  isFinal: false,
  hasError: false,
};

// Derive values for stateSites and domainSites in state. This is a one-time mapping we
// generate when sites are loaded into state containing lists of site codes for each
// state code / domain code.
const deriveRegionSites = (state) => {
  const stateSites = {};
  const domainSites = {};
  Object.keys(state.data.sites).forEach((siteCode) => {
    const { stateCode, domainCode } = state.data.sites[siteCode];
    if (!stateSites[stateCode]) { stateSites[stateCode] = new Set(); }
    if (!domainSites[domainCode]) { domainSites[domainCode] = new Set(); }
    stateSites[stateCode].add(siteCode);
    domainSites[domainCode].add(siteCode);
  });
  // Fill in empty sets for any states that had no NEON sites
  Object.keys(state.data.states).forEach((stateCode) => {
    if (!stateSites[stateCode]) { stateSites[stateCode] = new Set(); }
  });
  return {
    ...state,
    data: { ...state.data, stateSites, domainSites },
  };
};

/**
   CONTEXT
*/
const Context = createContext(DEFAULT_STATE);

/**
   HOOK
*/
const useNeonContextState = () => {
  const hookResponse = useContext(Context);
  // If called by a component that's not inside a NeonContext the hookResponse will be
  // DEFAULT_STATE. Otherwise it will return an array of length one containing the current state.
  // Thus we double-check here that we got an active state before returning it.
  if (!Array.isArray(hookResponse) || !hookResponse[0].isActive) {
    return [
      { ...DEFAULT_STATE },
      () => {},
    ];
  }
  return hookResponse;
};

/**
   Reducer
*/
const reducer = (state, action) => {
  // Always deep clone fetches as that's the main thing we care about
  // changing to trigger re-renders in the consumer.
  const newState = { ...state, fetches: cloneDeep(state.fetches) };
  const hasValidRemoteAsset = action.asset && Object.keys(REMOTE_ASSETS).includes(action.asset);
  switch (action.type) {
    case 'fetchCalled':
      if (!action.key || !state.fetches[action.key]) { return state; }
      newState.fetches[action.key].status = FETCH_STATUS.FETCHING;
      return newState;

    // Actions for handling sites fetch
    case 'fetchSitesSucceeded':
      newState.fetches.sites.status = FETCH_STATUS.SUCCESS;
      newState.data.sites = action.sites;
      newState.isFinal = true;
      return deriveRegionSites(newState);
    case 'fetchSitesFailed':
      newState.fetches.sites.status = FETCH_STATUS.ERROR;
      newState.fetches.sites.error = action.error;
      newState.isFinal = true;
      newState.hasError = true;
      return newState;

    // Actions for handling auth fetch
    case 'setAuthenticated':
      newState.auth.isAuthenticated = action.isAuthenticated;
      return newState;
    case 'setAuthWorking':
      newState.auth.isAuthWorking = action.isAuthWorking;
      return newState;
    case 'setAuthWsConnected':
      newState.auth.isAuthWsConnected = action.isAuthWsConnected;
      return newState;
    case 'fetchAuthSucceeded':
      newState.fetches.auth.status = FETCH_STATUS.SUCCESS;
      newState.auth.isAuthenticated = !!action.isAuthenticated;
      newState.auth.userData = AuthService.parseUserData(action.response);
      return newState;
    case 'fetchAuthFailed':
      newState.fetches.auth.status = FETCH_STATUS.ERROR;
      newState.auth.isAuthenticated = false;
      newState.auth.userData = null;
      return newState;

    // Actions for handling remote assets
    case 'fetchHtmlSucceeded':
      if (!hasValidRemoteAsset) { return state; }
      newState.fetches[action.asset].status = FETCH_STATUS.SUCCESS;
      newState.html[action.asset] = action.html;
      return newState;
    case 'fetchHtmlFailed':
      if (!hasValidRemoteAsset) { return state; }
      newState.fetches[action.asset].status = FETCH_STATUS.ERROR;
      newState.fetches[action.asset].error = action.error;
      return newState;

    default:
      return state;
  }
};

/**
   Function to convert sites fetch response to the shape we expect
*/
const parseSitesFetchResponse = (sitesArray = []) => {
  if (!Array.isArray(sitesArray)) { return {}; }
  const sitesObj = {};
  sitesArray.forEach((site) => {
    sitesObj[site.siteCode] = {
      siteCode: site.siteCode || site.code,
      description: site.siteDescription || site.description,
      type: site.siteType || site.type,
      stateCode: site.stateCode,
      domainCode: site.domainCode,
      latitude: site.siteLatitude || site.latitude,
      longitude: site.siteLongitude || site.longitude,
      terrain: site.terrain || sitesJSON[site.siteCode].terrain,
      zoom: site.zoom || sitesJSON[site.siteCode].zoom,
    };
  });
  return sitesObj;
};

/**
   Context Provider
*/
const Provider = (props) => {
  const { useCoreAuth, useCoreHeader, children } = props;

  const initialState = cloneDeep(DEFAULT_STATE);
  initialState.isActive = true;
  if (useCoreAuth) {
    initialState.auth.useCore = true;
    initialState.fetches.auth.status = FETCH_STATUS.AWAITING_CALL;
  }
  if (!useCoreHeader) {
    initialState.fetches[DRUPAL_HEADER_HTML].status = FETCH_STATUS.AWAITING_CALL;
    initialState.fetches[DRUPAL_FOOTER_HTML].status = FETCH_STATUS.AWAITING_CALL;
  }
  const [state, dispatch] = useReducer(reducer, initialState);

  // Method to sanitize partial HTML. As delivered presently there are some markup issues that
  // throw warnings when parsed with HTMLReactParser.
  const sanitizePartialHTML = html => html.replace(/ value=""/g, ' initialValue=""');

  // Method to fetch header and/or footer partials
  const fetchPartialHTML = (key) => {
    if (!Object.keys(REMOTE_ASSETS).includes(key)) { return; }
    const { url } = REMOTE_ASSETS[key];
    window.fetch(url, { method: 'GET', headers: { Accept: 'text/html' } })
      .then((res) => {
        if (res.status >= 400) { throw new Error(`${res.status} ${res.statusText}`); }
        return res.text();
      })
      .then((html) => {
        dispatch({ type: 'fetchHtmlSucceeded', asset: key, html: sanitizePartialHTML(html) });
        return of(true);
      })
      .catch((error) => {
        dispatch({ type: 'fetchHtmlFailed', asset: key, error });
        return of(false);
      });
  };

  // Identify any cascading authentication fetches that require
  // the WS to be connected to initiate.
  const cascadeAuthFetches = [];

  // Subject and effect to perform and manage the sites GraphQL fetch
  const fetchMethods = {
    sites: () => {
      NeonGraphQL.getAllSites().pipe(
        map((response) => {
          if (response.response && response.response.data && response.response.data.sites) {
            const sites = parseSitesFetchResponse(response.response.data.sites);
            dispatch({ type: 'fetchSitesSucceeded', sites });
            return of(true);
          }
          dispatch({ type: 'fetchSitesFailed', error: 'malformed response' });
          return of(false);
        }),
        catchError((error) => {
          dispatch({ type: 'fetchSitesFailed', error: error.message });
          return of(false);
        }),
      ).subscribe();
    },
    auth: () => {
      AuthService.fetchUserInfo(
        (response) => {
          const isAuthenticated = AuthService.isAuthenticated(response);
          if (!isAuthenticated && AuthService.isSsoLogin(response)) {
            // If we're not authenticated and have identified another SSO
            // application that's authenticated, trigger a silent authentication
            // check flow.
            if (AuthService.allowSilentAuth()) {
              if (!state.auth.isAuthWsConnected) {
                cascadeAuthFetches.push(() => AuthService.loginSilently(dispatch, true));
              } else {
                AuthService.loginSilently(dispatch, true);
              }
            } else {
              AuthService.login();
            }
          } else {
            dispatch({ type: 'fetchAuthSucceeded', isAuthenticated, response });
          }
          // Initialize a subscription to the auth WS
          AuthService.watchAuth0(dispatch, cascadeAuthFetches);
        },
        (error) => {
          dispatch({ type: 'fetchAuthFailed', error });
          // Initialize a subscription to the auth WS
          AuthService.watchAuth0(dispatch, cascadeAuthFetches);
        },
      );
    },
    [DRUPAL_HEADER_HTML]: () => fetchPartialHTML(DRUPAL_HEADER_HTML),
    [DRUPAL_FOOTER_HTML]: () => fetchPartialHTML(DRUPAL_FOOTER_HTML),
  };

  // Effect: Trigger all fetches that are awaiting call
  useEffect(() => {
    Object.keys(state.fetches).forEach((key) => {
      if (
        state.fetches[key].status !== FETCH_STATUS.AWAITING_CALL
          || typeof fetchMethods[key] !== 'function'
      ) { return; }
      dispatch({ type: 'fetchCalled', key });
      fetchMethods[key]();
    });
  });

  /**
     Render
  */
  return (
    <Context.Provider value={[state, dispatch]}>
      {children}
    </Context.Provider>
  );
};

Provider.propTypes = {
  useCoreAuth: PropTypes.bool,
  useCoreHeader: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.string,
    ])),
    PropTypes.node,
    PropTypes.string,
  ]).isRequired,
};

Provider.defaultProps = {
  useCoreAuth: false,
  useCoreHeader: false,
};

/**
   getWrappedComponent
*/
const getWrappedComponent = Component => (props) => {
  const [{ isActive }] = useNeonContextState();
  if (!isActive) {
    return (
      <Provider>
        <Component {...props} />
      </Provider>
    );
  }
  return <Component {...props} />;
};

/**
   Export
*/
const NeonContext = {
  Provider,
  useNeonContextState,
  DEFAULT_STATE,
  getWrappedComponent,
};
export default NeonContext;
