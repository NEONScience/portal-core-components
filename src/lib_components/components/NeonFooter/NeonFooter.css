.neon-footer {
  display: grid;
  grid-gap: 2px;
  grid-template-columns: minmax(330px, 100%) minmax(325px, 325px) minmax(
      200px,
      200px
    );
  grid-template-areas: "social contact neonlogo" "nsftext battelletext neonlogo";
  padding-top: 1em;
}

.neon-footer__box {
  font-size: 100%;
}

.neon-footer__div--social {
  float:left;
  grid-area: social;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  height: 50px;
  border-bottom: 1px rgba(208,208,208,0.6) solid;
  box-sizing: content-box;
}

.neon-footer__neon-logo {
  background-image: url("./images/social--newsletter.png");
  background-repeat: no-repeat;

  text-decoration-line: none;
  word-wrap: break-word;
  font-size: 12px;
  font-weight: 500;
  line-height: 14px;
  color: #00a1b1;
  max-width: 99px;
  padding-left: 35px;
}

.neon-footer__neon-logo:hover {
  background-image: url("./images/social--newsletter--hover.png");
  background-repeat: no-repeat;
}

.neon-footer__facebook {
  background-image: url("./images/social--facebook.png");
  background-repeat: no-repeat;
  width: 29px;
  text-indent: -9999px;
}

.neon-footer__facebook:hover {
  background-image: url("./images/social--facebook--hover.png");
  background-repeat: no-repeat;
}

.neon-footer__twitter {
  background-image: url("./images/social--twitter.png");
  background-repeat: no-repeat;
  width: 29px;
  text-indent: -9999px;
}

.neon-footer__twitter:hover {
  background-image: url("./images/social--twitter--hover.png");
  background-repeat: no-repeat;
}

.neon-footer__linkedin {
  background-image: url("./images/social--linkedin.png");
  background-repeat: no-repeat;
  width: 29px;
  text-indent: -9999px;
}
.neon-footer__linkedin:hover {
  background-image: url("./images/social--linkedin--hover.png");
  background-repeat: no-repeat;
}

.neon-footer__youtube {
  background-image: url("./images/social--youtube.png");
  background-repeat: no-repeat;
  width: 29px;
  text-indent: -9999px;
}

.neon-footer__youtube:hover {
  background-image: url("./images/social--youtube--hover.png");
  background-repeat: no-repeat;
}

.neon-footer__logo--social {
  min-height: 29px;
  /* padding-top: 5px; */
  padding-right: 6px;
}

.neon-footer__div--contact {
  float: right;
  grid-area: contact;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  height: 50px;
  border-bottom: 1px rgba(208,208,208,0.6) solid;
}

.neon-footer__link--contact:hover {
  text-decoration: underline;
}

.neon-footer__link--contact {
  min-height: 29px;
  padding-right: 6px;
  padding-top: 10px;
  margin-left: 4px;
  margin-right: 4px;
  color: #00a1b1;
  font-size: 12px;
  font-weight: 400;
  line-height: 18.36px;
  text-decoration-line: none;
  white-space: nowrap;
}

.neon-footer__icon--tower {
  width: 40px;
  background-image: url("./images/tower.png");
  background-repeat: no-repeat;
}

.neon-footer__text--nsf {
  float: left;
  grid-area: nsftext;
  color: #999;
  font-size: 12px;
  font-weight: 500;
  line-height: 18px;
  padding-top: 6.75px;
  padding-bottom: 6.75px;
  text-rendering: optimizeLegibility;
}

.neon-footer__span--nsf-text {
  max-width: 395px;
}

.neon-footer__text--battelle {
  float:right;
  grid-area: battelletext;
  color: #999;
  font-size: 12px;
  font-weight: 500;
  line-height: 18px;
  padding-top: 6.75px;
  padding-bottom: 6.75px;
  padding-right: 10px;
  text-rendering: optimizeLegibility;
  text-align: right;
}

.neon-footer__logo--footer{
  float: right;
  top: -0.75em;
  width: 11.25em;
}
