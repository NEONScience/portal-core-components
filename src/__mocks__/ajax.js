/**
   Ajax Mock

   Tests for any components using rjsx/ajax.getJSON can use this to inject a synchronous response.

   Usage:

   import { mockAjaxResponse, mockAjaxError } from 'path/to/src/__mocks__/ajax';

   // Set a mock response (for ajax() AND ajax.getJSON() methods)
   mockAjaxResponse({ ...response });

   // Set a mock error (for ajax() AND ajax.getJSON() methods)
   mockAjaxError('fail');
*/

import { of, throwError } from 'rxjs';

import { ajax } from 'rxjs/ajax';

jest.mock('rxjs/ajax');

ajax.mockReturnValue(of({}));

ajax.getJSON = jest.fn();
ajax.getJSON.mockReturnValue(of({}));

export function mockRawAjaxResponse(response = {}) {
  ajax.mockReset();
  ajax.mockReturnValue(of(response));
}

export function mockRawAjaxError(error = '') {
  ajax.mockReset();
  ajax.mockImplementation(() => throwError(new Error(error)));
}

export function mockGetJSONAjaxResponse(response = {}) {
  ajax.getJSON.mockReset();
  ajax.getJSON.mockReturnValue(of(response));
}

export function mockGetJSONAjaxError(error = '') {
  ajax.getJSON.mockReset();
  ajax.getJSON.mockImplementation(() => throwError(new Error(error)));
}

export function mockAjaxResponse(response = {}) {
  mockRawAjaxResponse(response);
  mockGetJSONAjaxResponse(response);
}

export function mockAjaxError(error = '') {
  mockRawAjaxError(error);
  mockGetJSONAjaxError(error);
}
