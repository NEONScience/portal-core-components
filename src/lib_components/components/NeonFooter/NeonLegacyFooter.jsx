import React from 'react';
import './NeonFooter.css';

import { makeStyles } from '@material-ui/core/styles';
import Divider from '@material-ui/core/Divider';
import Hidden from '@material-ui/core/Hidden';
import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import Container from '@material-ui/core/Container';
import Typography from '@material-ui/core/Typography';
import Theme from '../Theme/Theme';

import LogoFooter from './images/logo--footer.png';
import NewsletterIcon from './images/social--newsletter.png';

require('font-awesome/css/font-awesome.min.css');

const useStyles = makeStyles((theme) => ({
  box: {
    padding: theme.spacing(0),
    margin: theme.spacing(0),
  },
  container: {
    padding: theme.spacing(1, 4),
    backgroundColor: '#fff',
    boxShadow: '0 50vh 0 50vh #fff',
  },
  smContainer: {
    padding: theme.spacing(2, 0),
    display: 'flex',
    justifyContent: 'space-around',
  },
  smCaption: {
    padding: theme.spacing(0),
    marginBottom: theme.spacing(2),
    color: theme.palette.grey[300],
    textAlign: 'justify',
  },
  smCopyright: {
    padding: theme.spacing(0),
    color: theme.palette.grey[300],
    display: 'flex',
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  linkIcon: {
    color: theme.palette.grey[200],
    '&:hover, &:focus': {
      color: theme.palette.grey[500],
    },
  },
  newsletterIcon: {
    verticalAlign: 'middle',
    marginRight: theme.spacing(1),
    opacity: 0.6,
    '&:hover, &:focus': {
      opacity: 1,
    },
  },
  footerTextLink: {
    color: '#00a1b1',
  },
}));

export default function NeonFooter() {
  const classes = useStyles(Theme);
  const year = new Date().getFullYear();
  const copyrightDisplay = `Â© ${year.toString()} Battelle`;
  const bottomCaption = `
The National Ecological Observatory Network is a major facility fully funded by
the National Science Foundation. Any opinions, findings and conclusions or
recommendations expressed in this material do not necessarily reflect the views
of the National Science Foundation.
`;
  return (
    <div className={classes.box} data-selenium="neon-footer">
      <Divider />
      <Container className={classes.container}>
        <Hidden mdUp>
          <Grid container spacing={2}>
            <Grid item xs={6}>
              <Container style={{ padding: Theme.spacing(2, 0) }}>
                <Link
                  className={classes.footerTextLink}
                  title="Sign up for the NEON Newsletter"
                  href="http://visitor.r20.constantcontact.com/manage/optin?v=001VGGvr8xQIKjYc_kFiPLlwv2eCR5mvCy-l84uOGp_mC-N67a0MgiOjSJIWa1ZibyTJYm9YZbvasF-RX5OWfNMI7gIq2IjIF6fDARps7B---g%3D"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <img
                    title="Sign up for the NEON Newsletter"
                    alt="Sign up for the NEON Newsletter"
                    src={NewsletterIcon}
                    className={classes.newsletterIcon}
                  />
                  Newsletter
                </Link>
              </Container>
            </Grid>
            <Grid item xs={6}>
              <Container className={classes.smContainer}>
                <Link
                  className={classes.linkIcon}
                  title="Facebook"
                  href="https://www.facebook.com/NEONScienceData"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <i className="fa fa-facebook-square fa-2x fa-fw" />
                </Link>
                <Link
                  className={classes.linkIcon}
                  title="Twitter"
                  href="https://twitter.com/NEON_Sci"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <i className="fa fa-twitter-square fa-2x fa-fw" />
                </Link>
                <Link
                  className={classes.linkIcon}
                  title="LinkedIn"
                  href="https://www.linkedin.com/company/neon-science/"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <i className="fa fa-linkedin-square fa-2x fa-fw" />
                </Link>
                <Link
                  className={classes.linkIcon}
                  title="YouTube"
                  href="https://www.youtube.com/neonscience"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <i className="fa fa-youtube-square fa-2x fa-fw" />
                </Link>
              </Container>
            </Grid>
          </Grid>
          <Divider />
          <Container className={classes.smContainer}>
            <Link className={classes.footerTextLink} href="http://www.neonscience.org/about/contact">Contact</Link>
            <Link className={classes.footerTextLink} href="http://www.neonscience.org/opportunities/careers">Careers</Link>
            <Link className={classes.footerTextLink} href="http://www.neonscience.org/about/faq">FAQ</Link>
            <Link className={classes.footerTextLink} href="http://www.neonscience.org/about/contact/media">Media</Link>
            <Link className={classes.footerTextLink} href="http://www.neonscience.org/about/terms-use">Terms of Use</Link>
          </Container>
          <Container className={classes.smCaption}>
            <Typography variant="caption">
              {bottomCaption}
            </Typography>
          </Container>
          <Container className={classes.smCopyright}>
            <Typography variant="caption">
              {copyrightDisplay}
            </Typography>
            <img
              src={LogoFooter}
              style={{ maxHeight: Theme.spacing(2) }}
              title="Proudly operated by Battelle"
              alt="Proudly operated by Battelle"
            />
          </Container>
        </Hidden>
        <Hidden smDown>
          <div className="neon-footer">
            <img
              src={LogoFooter}
              className="neon-footer__logo--footer"
              title="Proudly operated by Battelle"
              alt="Proudly operated by Battelle"
            />
            <div className="neon-footer__box neon-footer__div--social">
              <a
                title="Sign up for our Newsletter"
                className="neon-footer__social--logo neon-footer__neon-logo"
                href="http://visitor.r20.constantcontact.com/manage/optin?v=001VGGvr8xQIKjYc_kFiPLlwv2eCR5mvCy-l84uOGp_mC-N67a0MgiOjSJIWa1ZibyTJYm9YZbvasF-RX5OWfNMI7gIq2IjIF6fDARps7B---g%3D"
                target="_blank"
                rel="noopener noreferrer"
              >
                Sign Up for Our Newsletter
              </a>

              <a
                title="Facebook"
                className="neon-footer__social--logo neon-footer__facebook"
                href="https://www.facebook.com/NEONScienceData"
                target="_blank"
                rel="noopener noreferrer"
              >
                Facebook
              </a>

              <a
                title="Twitter"
                className="neon-footer__social--logo neon-footer__twitter"
                href="https://twitter.com/NEON_Sci"
                target="_blank"
                rel="noopener noreferrer"
              >
                Twitter
              </a>

              <a
                title="LinkedIn"
                className="neon-footer__social--logo neon-footer__linkedin"
                href="https://www.linkedin.com/company/neon-science/"
                target="_blank"
                rel="noopener noreferrer"
              >
                LinkedIn
              </a>

              <a
                title="YouTube"
                className="neon-footer__logo--social neon-footer__youtube"
                href="https://www.youtube.com/neonscience"
                target="_blank"
                rel="noopener noreferrer"
              >
                YouTube
              </a>
            </div>

            <div className="neon-footer__box neon-footer__div--contact">
              <a
                className="neon-footer__link--contact"
                href="http://www.neonscience.org/about/contact"
              >
                Contact
              </a>
              <a
                className="neon-footer__link--contact"
                href="http://www.neonscience.org/opportunities/careers"
              >
                Careers
              </a>
              <a
                className="neon-footer__link--contact"
                href="http://www.neonscience.org/about/faq"
              >
                FAQ
              </a>
              <a
                className="neon-footer__link--contact"
                href="http://www.neonscience.org/about/contact/media"
              >
                Media
              </a>
              <a
                className="neon-footer__link--contact"
                href="http://www.neonscience.org/about/terms-use"
              >
                Terms of Use
              </a>
              <div className="neon-footer__icon--tower" />
            </div>

            <div className="neon-footer__box neon-footer__text--nsf">
              <div className="neon-footer__span--nsf-text">
                {bottomCaption}
              </div>
            </div>

            <div className="neon-footer__box neon-footer__text--battelle">{copyrightDisplay}</div>

          </div>
        </Hidden>
      </Container>
    </div>
  );
}
