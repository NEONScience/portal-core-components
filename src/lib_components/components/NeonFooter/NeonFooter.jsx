import React from 'react';
import PropTypes from 'prop-types';
import HTMLReactParser from 'html-react-parser';

import Skeleton from '@material-ui/lab/Skeleton';

import NeonContext, { FETCH_STATUS } from '../NeonContext/NeonContext';

import NeonLegacyFooter from './NeonLegacyFooter';

const NeonFooter = (props) => {
  const { drupalCssLoaded } = props;
  const [{
    isActive,
    fetches: { footer: footerFetch },
    html: { footer: footerHTML },
  }] = NeonContext.useNeonContextState();

  let renderMode = 'legacy';
  if (isActive) {
    if (footerFetch.status === FETCH_STATUS.SUCCESS && footerHTML && drupalCssLoaded) {
      renderMode = 'drupal';
    }
    if ([FETCH_STATUS.AWAITING_CALL, FETCH_STATUS.FETCHING].includes(footerFetch.status)) {
      renderMode = 'loading';
    }
  }

  switch (renderMode) {
    case 'loading':
      return (
        <footer id="footer">
          <Skeleton variant="rect" height="300px" width="100%" />
        </footer>
      );

    case 'drupal':
      return (
        <footer id="footer">
          {HTMLReactParser(footerHTML)}
        </footer>
      );

    default:
      return <NeonLegacyFooter />;
  }
};

NeonFooter.propTypes = {
  drupalCssLoaded: PropTypes.bool,
};

NeonFooter.defaultProps = {
  drupalCssLoaded: false,
};

export default NeonFooter;
