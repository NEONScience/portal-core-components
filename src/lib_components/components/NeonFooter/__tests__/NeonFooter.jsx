import React from 'react';
import renderer from 'react-test-renderer';

import '../../../../__mocks__/NeonContext';
import NeonContext, { FETCH_STATUS } from '../../NeonContext/NeonContext';

import NeonFooter from '../NeonFooter';

import REMOTE_ASSETS from '../../../remoteAssetsMap/remoteAssetsMap';

jest.mock('../../../remoteAssets/drupal-footer.html', () => (
  '<div>MOCK-FALLBACK-DRUPAL-FOOTER-HTML</div>'
));
const DRUPAL_FOOTER_HTML = REMOTE_ASSETS.DRUPAL_FOOTER_HTML.KEY;

describe('NeonFooter', () => {
  beforeEach(() => {
    NeonContext.useNeonContextState.mockReset();
  });
  test('renders with no props and inactive NeonContext state', () => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: false,
      fetches: { [DRUPAL_FOOTER_HTML]: { status: null } },
      html: { [DRUPAL_FOOTER_HTML]: null },
    }]);
    const tree = renderer
      .create(<NeonFooter />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders with no props and active/fetching NeonContext state', () => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: true,
      fetches: { [DRUPAL_FOOTER_HTML]: { status: FETCH_STATUS.FETCHING } },
      html: { [DRUPAL_FOOTER_HTML]: null },
    }]);
    const tree = renderer
      .create(<NeonFooter />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders with no props and error NeonContext state', () => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: true,
      fetches: { [DRUPAL_FOOTER_HTML]: { status: FETCH_STATUS.ERROR } },
      html: { [DRUPAL_FOOTER_HTML]: null },
    }]);
    const tree = renderer
      .create(<NeonFooter />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders with no props and success NeonContext state', () => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: true,
      fetches: { [DRUPAL_FOOTER_HTML]: { status: FETCH_STATUS.SUCCESS } },
      html: { [DRUPAL_FOOTER_HTML]: '<div>test drupal html</div>' },
    }]);
    const tree = renderer
      .create(<NeonFooter />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders fallback with drupalCSSLoaded prop and error NeonContext state', () => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: true,
      fetches: { [DRUPAL_FOOTER_HTML]: { status: FETCH_STATUS.ERROR } },
      html: { [DRUPAL_FOOTER_HTML]: null },
    }]);
    const tree = renderer
      .create(<NeonFooter drupalCssLoaded />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders with drupalCssLoaded prop and success NeonContext state', () => {
    NeonContext.useNeonContextState.mockReturnValue([{
      isActive: true,
      fetches: { [DRUPAL_FOOTER_HTML]: { status: FETCH_STATUS.SUCCESS } },
      html: { [DRUPAL_FOOTER_HTML]: '<div>test drupal html</div>' },
    }]);
    const tree = renderer
      .create(<NeonFooter drupalCssLoaded />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
