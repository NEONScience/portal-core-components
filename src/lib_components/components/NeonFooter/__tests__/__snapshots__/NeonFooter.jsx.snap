// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeonFooter renders fallback with drupalCSSLoaded prop and error NeonContext state 1`] = `
<footer
  id="footer"
>
  <div>
    MOCK-FALLBACK-DRUPAL-FOOTER-HTML
  </div>
</footer>
`;

exports[`NeonFooter renders with drupalCssLoaded prop and success NeonContext state 1`] = `
<footer
  id="footer"
>
  <div>
    test drupal html
  </div>
</footer>
`;

exports[`NeonFooter renders with no props and active/fetching NeonContext state 1`] = `
<footer
  id="footer"
>
  <span
    className="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
    style={
      Object {
        "height": "300px",
        "width": "100%",
      }
    }
  />
</footer>
`;

exports[`NeonFooter renders with no props and error NeonContext state 1`] = `
<footer
  id="footer"
>
  <span
    className="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
    style={
      Object {
        "height": "300px",
        "width": "100%",
      }
    }
  />
</footer>
`;

exports[`NeonFooter renders with no props and inactive NeonContext state 1`] = `
<div
  className="makeStyles-box-1"
  data-selenium="neon-footer"
>
  <hr
    className="MuiDivider-root"
  />
  <div
    className="MuiContainer-root makeStyles-container-2 MuiContainer-maxWidthLg"
  />
</div>
`;

exports[`NeonFooter renders with no props and success NeonContext state 1`] = `
<footer
  id="footer"
>
  <span
    className="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
    style={
      Object {
        "height": "300px",
        "width": "100%",
      }
    }
  />
</footer>
`;
